conn_go18.go

```diff
@@ -49,41 +51,51 @@ func (cn *conn) BeginTx(ctx context.Context, opts driver.TxOptions) (driver.Tx,
 	if err != nil {
 		return nil, err
 	}
-	cn.txnClosed = cn.watchCancel(ctx)
+	cn.txnFinish = cn.watchCancel(ctx)
 	return tx, nil
 }
 
-func (cn *conn) watchCancel(ctx context.Context) chan<- struct{} {
+func (cn *conn) watchCancel(ctx context.Context) func() {
 	if done := ctx.Done(); done != nil {
-		closed := make(chan struct{})
+		finished := make(chan struct{})
 		go func() {
 			select {
 			case <-done:
-				cn.cancel()
-			case <-closed:
+				_ = cn.cancel()
+				finished <- struct{}{}
+			case <-finished:
 			}
 		}()
-		return closed
+		return func() {
+			select {
+			case <-finished:
+			case finished <- struct{}{}:
+			}
+		}
 	}
 	return nil
 }
 
-func (cn *conn) cancel() {
-	var err error
-	can := &conn{}
-	can.c, err = dial(cn.dialer, cn.opts)
+func (cn *conn) cancel() error {
+	c, err := dial(cn.dialer, cn.opts)
 	if err != nil {
-		return
+		return err
+	}
+	can := conn{
+		c: c,
 	}
 	can.ssl(cn.opts)
 
-	defer can.errRecover(&err)
-
 	w := can.writeBuf(0)
 	w.int32(80877102) // cancel request code
 	w.int32(cn.processID)
 	w.int32(cn.secretKey)
 
 	can.sendStartupPacket(w)
-	_ = can.c.Close()
+	can.sendStartupPacket(w)
```

Comment:
sendStartupPacket panics on various errors: https://github.com/lib/pq/blob/e4af84aab01e159ca479940dbde304519e8e8811/conn.go#L896

I'm not sure removing the errRecover is safe.